File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    // worst case complexity -> every element will put in the queue 
    // for each element of queue there is -> k is average word size and there 26 iterations so total = k*26
    // O(n*k*26) = O(n*k) but there is set also we can remove all elements from set
    // O(n*k*26 + nlogn) 
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        // brute force
        set<string>s;
        for(auto word : wordList)s.insert(word);
        s.erase(beginWord);

        queue<pair<string,int>>q;
        q.push({beginWord,1});
        int count = 0;

        while(!q.empty()){
            pair<string,int> p = q.front();
            q.pop();

            string word = p.first;
            int level = p.second;

            if(word == endWord)return level;
            
            for(int i = 0; i < word.size(); i++){
                string dummyWord = word;
                for(int j = 0; j < 26; j++){
                    char ch = 'a'+j;
                    dummyWord[i] = ch;

                    if(s.find(dummyWord) != s.end()){
                        s.erase(dummyWord);
                        q.push({dummyWord,level+1});
                    }
                }
            }
        }

        return 0;
    }
};

File: 0127-word-ladder/0127-word-ladder.java
class Pair<K, V> {
    private final K key;
    private final V value;

    public Pair(K key, V value) {
        this.key = key;
        this.value = value;
    }

    public K getKey() {
        return key;
    }

    public V getValue() {
        return value;
    }
}
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        HashMap<String, Integer> m = new HashMap<>();
        for (String word : wordList) {
            m.put(word, m.getOrDefault(word, 0) + 1);
        }

        Queue<Pair<String,Integer>>q = new LinkedList<>();
        q.add(new Pair<>(beginWord,1));
        m.remove(beginWord);

        while(!q.isEmpty()){
            Pair<String,Integer>p = q.remove();
            String word = p.getKey();
            int level = p.getValue();

            if(word.equals(endWord)){
                return level;
            }

            for (int i = 0; i < word.length(); i++) {
                String dummyWord = word;
                for (char ch = 'a'; ch <= 'z'; ch++) {
                    if (word.charAt(i) == ch) continue;

                    StringBuilder sb = new StringBuilder(dummyWord);
                    sb.setCharAt(i, ch);
                    String str = sb.toString();

                    if (m.containsKey(str)) {
                        q.add(new Pair<>(str, level + 1)); 
                        m.remove(str);
                    }
                }
            }

        }

        return 0;
    }
}